[SPARK-8230][SQL] Add array/map size method

EntilZha opened the pull request:
Pull Request for: https://issues.apache.org/jira/browse/SPARK-8230

Primary issue resolved is to implement array/map size for Spark SQL. Code is ready for review by a committer. Chen Hao is on the JIRA ticket, but I don't know his username on github, @rxin is also on JIRA ticket.

Things to review:
1. Where to put added functions namespace wise, they seem to be part of a few operations on collections which includes `sort_array` and `array_contains`. Hence the name given `collectionOperations.scala` and `_collection_functions` in python.
2. In Python code, should it be in a `1.5.0` function array or in a collections array?
3. Are there any missing methods on the `Size` case class? Looks like many of these functions have generated Java code, is that also needed in this case?
4. Something else?

Commit:
First attempt at implementing size for maps and arrays

Commit:
First attempt at implementing size for maps and arrays

EntilZha committed:
Finished working version of size function and added it to python

EntilZha committed:
Finished working version of size function and added it to python

EntilZha committed:
added unit test for size

EntilZha committed:
added unit test for size

EntilZha committed:
added documentation

EntilZha committed:
added unit tests for map size

EntilZha committed:
added unit tests for map size

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Given an array or map, returns its size.
+ */
+case class Size(child: Expression) extends UnaryExpression {
```
extends ExpectsInputTypes

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -0,0 +1,46 @@
+
+  override def dataType: DataType = IntegerType
+
+  override def checkInputDataTypes(): TypeCheckResult = {
```
once you extends ExpectsInputTypes, you can use a TypeCollection to specify the type

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -0,0 +1,46 @@
+      case ArrayType(_, _) => child.eval(input).asInstanceOf[Seq[Any]].size
+      case MapType(_, _, _) => child.eval(input).asInstanceOf[Map[Any, Any]].size
+    }
+  }
```
can you add a codegen version?

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/functions.scala:
```diff
@@ -2186,6 +2186,24 @@ object functions {
+
+  /**
+   * Returns length of array or map
+   * @group collection_funcs
```
i think you need to update the doc in the beginning of this file to define this new group

rxin commented:
We should also add some expression unit test. Take a look at the expressions package in the test folder and see how other things are done.

Thanks a lot for working on this.

chenghao-intel commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -0,0 +1,46 @@
+    }
+  }
+
+  override def eval(input: InternalRow): Int = {
```
override the function `nullSafeEval` instead.

EntilZha commented:
Thanks @rxin for feedback. Will make those changes tomorrow and repost when PR is ready for review again.

SparkQA commented:
[Test build #37606 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37606/consoleFull) for   PR 7462 at commit [`6f4aa7d`](https://github.com/apache/spark/commit/6f4aa7d9768dd606022ce90b33f0477422611fbd).

SparkQA commented:
[Test build #37606 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37606/console) for   PR 7462 at commit [`6f4aa7d`](https://github.com/apache/spark/commit/6f4aa7d9768dd606022ce90b33f0477422611fbd).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Size(child: Expression) extends UnaryExpression `

AmplabJenkins commented:
Merged build finished. Test FAILed.

chenghao-intel commented:
And as #7214 (In StringFunctionsSuite.scala) shows, we need to add the test suite for the expression evaluation directly.

The function of `checkEvaluation` will be helpful for testing the codegen stuff.

EntilZha committed:
refactored code to use better abstract classes/traits and implementations

EntilZha committed:
refactored code to use better abstract classes/traits and implementations

tarekauel commented on python/pyspark/sql/functions.py:
```diff
@@ -652,6 +658,16 @@ def ntile(n):
+def size(col):
+    """
+    Collection function: returns the length of the array or map stored in the column.
+    :param col: name of column or expression
```
Could you add an example/test here?

EntilZha committed:
updated python code with docs, switched classes/traits implemented, added (failing) expression tests

EntilZha committed:
updated python code with docs, switched classes/traits implemented, added (failing) expression tests

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

EntilZha commented:
I pushed updates to the code based on the comments above. I wrote an expressions test, but can't figure out why its failing. I get the following stacktrace:
```
[info] CollectionFunctionsSuite:
[info] - Array and Map Size *** FAILED *** (958 milliseconds)
[info]   Exception evaluating size([I@7e14fccd) (ExpressionEvalHelper.scala:69)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:496)
[info]   at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1555)
[info]   at org.scalatest.Assertions$class.fail(Assertions.scala:1348)
[info]   at org.scalatest.FunSuite.fail(FunSuite.scala:1555)
[info]   at org.apache.spark.sql.catalyst.expressions.ExpressionEvalHelper$class.checkEvaluationWithoutCodegen(ExpressionEvalHelper.scala:69)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite.checkEvaluationWithoutCodegen(CollectionFunctionsSuite.scala:24)
[info]   at org.apache.spark.sql.catalyst.expressions.ExpressionEvalHelper$class.checkEvaluation(ExpressionEvalHelper.scala:43)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite.checkEvaluation(CollectionFunctionsSuite.scala:24)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite$$anonfun$1.apply$mcV$sp(CollectionFunctionsSuite.scala:31)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite$$anonfun$1.apply(CollectionFunctionsSuite.scala:26)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite$$anonfun$1.apply(CollectionFunctionsSuite.scala:26)
[info]   at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:166)
[info]   at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:42)
[info]   at org.scalatest.FunSuiteLike$class.invokeWithFixture$1(FunSuiteLike.scala:163)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.FunSuiteLike$class.runTest(FunSuiteLike.scala:175)
[info]   at org.scalatest.FunSuite.runTest(FunSuite.scala:1555)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
[info]   at scala.collection.immutable.List.foreach(List.scala:318)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:396)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
[info]   at org.scalatest.FunSuiteLike$class.runTests(FunSuiteLike.scala:208)
[info]   at org.scalatest.FunSuite.runTests(FunSuite.scala:1555)
[info]   at org.scalatest.Suite$class.run(Suite.scala:1424)
[info]   at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1555)
[info]   at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
[info]   at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:545)
[info]   at org.scalatest.FunSuiteLike$class.run(FunSuiteLike.scala:212)
[info]   at org.scalatest.FunSuite.run(FunSuite.scala:1555)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:462)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:671)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:294)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:284)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:262)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
[info]   at java.lang.Thread.run(Thread.java:745)
[info]   Cause: java.lang.ClassCastException: [I cannot be cast to scala.collection.Seq
[info]   at org.apache.spark.sql.catalyst.expressions.Size.nullSafeEval(collectionOperations.scala:30)
[info]   at org.apache.spark.sql.catalyst.expressions.Size.nullSafeEval(collectionOperations.scala:25)
[info]   at org.apache.spark.sql.catalyst.expressions.UnaryExpression.eval(Expression.scala:219)
[info]   at org.apache.spark.sql.catalyst.expressions.ExpressionEvalHelper$class.evaluate(ExpressionEvalHelper.scala:62)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite.evaluate(CollectionFunctionsSuite.scala:24)
[info]   at org.apache.spark.sql.catalyst.expressions.ExpressionEvalHelper$class.checkEvaluationWithoutCodegen(ExpressionEvalHelper.scala:69)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite.checkEvaluationWithoutCodegen(CollectionFunctionsSuite.scala:24)
[info]   at org.apache.spark.sql.catalyst.expressions.ExpressionEvalHelper$class.checkEvaluation(ExpressionEvalHelper.scala:43)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite.checkEvaluation(CollectionFunctionsSuite.scala:24)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite$$anonfun$1.apply$mcV$sp(CollectionFunctionsSuite.scala:31)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite$$anonfun$1.apply(CollectionFunctionsSuite.scala:26)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite$$anonfun$1.apply(CollectionFunctionsSuite.scala:26)
[info]   at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:166)
[info]   at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:42)
[info]   at org.scalatest.FunSuiteLike$class.invokeWithFixture$1(FunSuiteLike.scala:163)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.FunSuiteLike$class.runTest(FunSuiteLike.scala:175)
[info]   at org.scalatest.FunSuite.runTest(FunSuite.scala:1555)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
[info]   at scala.collection.immutable.List.foreach(List.scala:318)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:396)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
[info]   at org.scalatest.FunSuiteLike$class.runTests(FunSuiteLike.scala:208)
[info]   at org.scalatest.FunSuite.runTests(FunSuite.scala:1555)
[info]   at org.scalatest.Suite$class.run(Suite.scala:1424)
[info]   at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1555)
[info]   at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
[info]   at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:545)
[info]   at org.scalatest.FunSuiteLike$class.run(FunSuiteLike.scala:212)
[info]   at org.scalatest.FunSuite.run(FunSuite.scala:1555)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:462)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:671)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:294)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:284)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:262)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
[info]   at java.lang.Thread.run(Thread.java:745)
```

What I gather is basically Scala can't convert a class of type `I` (don't know what type this is, perhaps type erasure makes it no possible to know?) to `Seq` in `collectionOperations.scala:30`. I am not sure why this is.

What is the difference between tests in dataframes and tests in expressions (in terms of codepath)?

Other questions from changing code:
1. Difference between `eval` and `nullSafeEval`?
2. Quick summary of types to know for Spark SQL in terms of types of expressions (Unary, Binary, ExpectsInputTypes).
3. Does `Literal` represent Arrays/Maps, or only primitive types? Looks like you can either define your own datatype or have scala attempt to infer it with pattern matching.

SparkQA commented:
[Test build #37799 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37799/consoleFull) for   PR 7462 at commit [`f8a34d7`](https://github.com/apache/spark/commit/f8a34d7067a5eafd553af6c6fa48e010d66d01fe).

SparkQA commented:
[Test build #37799 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37799/console) for   PR 7462 at commit [`f8a34d7`](https://github.com/apache/spark/commit/f8a34d7067a5eafd553af6c6fa48e010d66d01fe).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `case class Size(child: Expression) extends UnaryExpression with ExpectsInputTypes `

AmplabJenkins commented:
Build finished. Test FAILed.

EntilZha committed:
fix style

EntilZha committed:
fix style

chenghao-intel commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionFunctionsSuite.scala:
```diff
@@ -0,0 +1,43 @@
+    val a1 = Literal.create(Array[Integer](), ArrayType(IntegerType))
+    val a2 = Literal.create(Array(1, 2), ArrayType(IntegerType))
+
+    checkEvaluation(Size(a0), 3)
```
As you did in https://github.com/apache/spark/pull/7462/files#diff-9853dcf5ce3d2ac1e94d473197ff5768R30
The ArrayType should be the `Seq[_]`, not `Array`.

chenghao-intel replied:
See Cause: java.lang.ClassCastException: [I cannot be cast to scala.collection.Seq

chenghao-intel replied:
`[I` means the type of `Array of Integer`.

EntilZha replied:
Is this spark representation of types or scala representation of types? Fixed that, fixing another error from codegen. When I get that figured out (or stuck) will push that up

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #37801 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37801/consoleFull) for   PR 7462 at commit [`4d0e29d`](https://github.com/apache/spark/commit/4d0e29d90dd092f6da4e9cffed21d10407c79664).

EntilZha committed:
fixed failing test

EntilZha committed:
fixed failing test

EntilZha commented:
Fixed tests using feedback from @chenghao-intel

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

tarekauel commented:
@EntilZha 
1. `eval` and `nullSafeEval`
`eval` will be invoked to evaluate the expression. Most expressions should return `null` if one of there arguments is `null`. In order to avoid that every expression has to check if `left` or `right` is `null`, `nullSafeEval` has been added. `eval` does the null check and calls `nullSafeEval`, see. 
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala#L289-L313
You should override `eval` if you don't want to return `null`, if one the arguments is `null`. Most of the times you will use `nullSafeEval`.
2. 
`UnaryExpression`: Expression has one parameter (like `size(x)`)

`BinaryExpression`: Expression has two parameters (like `contains(a, b)`)

`ExpectsInputTypes`: Allows to automatically check if the argument type is correct, see https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpectsInputTypes.scala#L42-L57. You specify the allowed types by overriding `inputTypes`.

`ImplicitCastInputs`: The difference to `ExpectsInputTypes` is that this tries to cast the value. Most string operations are implemented with a byte array as input. A string can be "casted" to a byte array by calling `.getBytes`. `ImplicitCastInputs` allows to call `contains(s: String, s2: String)` and `contains(s: Array[Byte], s2: Array[Byte])`. Typically you use this if a cast is reasonable. Cast from anything else to string is most of the times reasonable, but casting a string (automatically == implicit) to an integer value is most of the time not helpful. Users could still invoke the `cast` function. 
3. I don't know
4. Intellij allows to run most suites from the IDE. And have a look at https://cwiki.apache.org/confluence/display/SPARK/Useful+Developer+Tools#UsefulDeveloperTools-RunningIndividualTests

SparkQA commented:
[Test build #37802 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37802/consoleFull) for   PR 7462 at commit [`710780c`](https://github.com/apache/spark/commit/710780c42768f6a27b29f707366f0f476cbc0eac).

chenghao-intel commented:
> 1. Difference between eval and nullSafeEval?
`nullSafeEval` just to save some code for the null value checking.
```scala
  // from the BinaryExpression.eval
  override def eval(input: InternalRow): Any = {
    val value1 = left.eval(input)
    if (value1 == null) {
      null
    } else {
      val value2 = right.eval(input)
      if (value2 == null) {
        null
      } else {
        nullSafeEval(value1, value2)
      }
    }
  }
```
> 2. Quick summary of types to know for Spark SQL in terms of types of expressions (Unary, Binary, ExpectsInputTypes).
`Unary` means the expression only has a single child (e.g. Upper, Lower, Log ..).
`Binary' means the expression has 2 children (Add, Minus).
`ExpectsInputTypes` means the expression accept certain data types for its children. (e.g. Length only accept a child expression whose output type is `StringType` or `BinaryType`, otherwise we will throw runtime exception.
`ImplicitCastInputTypes` means the expression accept certain data types for its children. The difference with `ExpectsInputTypes` is, Spark SQL will auto convert the data types if data type of the children doesn't match. (e.g. StringSpace, which accept a IntegerType of child expression, and Spark SQL will auto convert it if a ShortType of child expression given).
> 3. Does Literal represent Arrays/Maps, or only primitive types? Looks like you can either define your own datatype or have scala attempt to infer it with pattern matching.
Literal can accept any type of data, not just the primitive types. But for the complex data type, you probably need to pass in the data type, as the type info were erased during the runtime, we may not able to infer the internal data type recursively.
> 4. How to run scalastyle properly configured to pass jenkins check
build/sbt -Phive -Phive-thriftserver scalastyle test:scalastyle

cloud-fan commented:
4. How to run scalastyle properly configured to pass jenkins check
./dev/scalastyle

SparkQA commented:
[Test build #37801 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37801/console) for   PR 7462 at commit [`4d0e29d`](https://github.com/apache/spark/commit/4d0e29d90dd092f6da4e9cffed21d10407c79664).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `case class Size(child: Expression) extends UnaryExpression with ExpectsInputTypes `

AmplabJenkins commented:
Build finished. Test FAILed.

SparkQA commented:
[Test build #37802 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37802/console) for   PR 7462 at commit [`710780c`](https://github.com/apache/spark/commit/710780c42768f6a27b29f707366f0f476cbc0eac).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `case class Size(child: Expression) extends UnaryExpression with ExpectsInputTypes `

AmplabJenkins commented:
Build finished. Test PASSed.

EntilZha commented:
Any other comments on code @rxin or @chenghao-intel? If it looks good, should I fix the merge conflict/rebase to current master or would you do that when you merge the PR?

rxin commented:
Please resolve the merge conflict if you can. Thanks.

rxin commented on python/pyspark/sql/functions.py:
```diff
@@ -196,6 +200,8 @@ def _():
     globals()[_name] = since(1.4)(_create_binary_mathfunction(_name, _doc))
 for _name, _doc in _window_functions.items():
     globals()[_name] = since(1.4)(_create_window_function(_name, _doc))
+for _name, _doc in _collection_functions.items():
```
i don't think you need this loop here

rxin commented on python/pyspark/sql/functions.py:
```diff
@@ -188,6 +188,10 @@ def _():
         This is equivalent to the PERCENT_RANK function in SQL.""",
 }
 
+_collection_functions = {
```
can you remove this, and just add size directly to __all__?

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -0,0 +1,43 @@
+    }
+  }
+
+  override def prettyName: String = "size"
```
you don't need this since prettyName's default implementation just takes the lowercase of the class name

Commit:
updated code to add size to __all__ directly and removed redundent pretty print

Commit:
fixed compile bug instroduced when merging

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

EntilZha commented:
@rxin, I fixed those python issues and rebased to master/resolved merge conflicts. Hopefully didn't break anything with git. Tests also should be passing. Anything else aside from waiting for Jenkins?

EntilZha commented:
Actually, I just looked at the diff, I did do something incorrect when I fixed merge conflicts/rebased, what should I do to fix that?

SparkQA commented:
[Test build #37856 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37856/consoleFull) for   PR 7462 at commit [`2eb10e4`](https://github.com/apache/spark/commit/2eb10e44370eac7e67966332114968479ac7a46c).

EntilZha commented:
Working on rebasing correctly, think I understand what I did/how to fix

Commit:
removed duplicate function from rebase/merge

Commit:
removed python code

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

EntilZha commented:
Merge/rebase is done, now ready for review

SparkQA commented:
[Test build #37861 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37861/consoleFull) for   PR 7462 at commit [`d88247c`](https://github.com/apache/spark/commit/d88247cfe02a1cc7411ea1111dee31a79b8e44bc).

SparkQA commented:
[Test build #37856 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37856/console) for   PR 7462 at commit [`2eb10e4`](https://github.com/apache/spark/commit/2eb10e44370eac7e67966332114968479ac7a46c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait Nondeterministic `
  * `case class Size(child: Expression) extends UnaryExpression with ExpectsInputTypes `
  * `abstract class RDG extends LeafExpression with Nondeterministic `
  * `case class Rand(seed: Long) extends RDG `
  * `case class Randn(seed: Long) extends RDG `

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #37861 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37861/console) for   PR 7462 at commit [`d88247c`](https://github.com/apache/spark/commit/d88247cfe02a1cc7411ea1111dee31a79b8e44bc).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait Nondeterministic `
  * `case class Size(child: Expression) extends UnaryExpression with ExpectsInputTypes `
  * `abstract class RDG extends LeafExpression with Nondeterministic `
  * `case class Rand(seed: Long) extends RDG `
  * `case class Randn(seed: Long) extends RDG `

AmplabJenkins commented:
Merged build finished. Test PASSed.

tarekauel commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -0,0 +1,40 @@
+  }
+
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
+    child.dataType match {
```
1. Pattern matching is not necessary here. Just do `defineCodeGen(ctx, ev, c => s"($c).size()")`.
2. Maybe we should call here `nullSafeCodeGen`  instead of `defineCodeGen`.

EntilZha replied:
With redefining `genCode` like this:

```scala
  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
    nullSafeCodeGen(ctx, ev, c => s"($c).size()")
  }
```

I am getting this error:
```scala
[info] - Array and Map Size *** FAILED *** (1 second, 503 milliseconds)
[info]   Code generation of size(List(1, 2, 3)) failed:
[info]
[info]         /* expression: List(1, 2, 3) */
[info]         Object obj4 = expressions[0].eval(i);
[info]         boolean isNull2 = obj4 == null;
[info]         scala.collection.Seq primitive3 = null;
[info]         if (!isNull2) {
[info]           primitive3 = (scala.collection.Seq) obj4;
[info]         }
[info]
[info]         boolean isNull0 = isNull2;
[info]         int primitive1 = -1;
[info]         if (!isNull0) {
[info]           (primitive3).size()
[info]         }
[info]
[info]   java.util.concurrent.ExecutionException: org.codehaus.commons.compiler.CompileException: Line 44, Column 27: Expression "(primitive3).size()" is not a type (ExpressionEvalHelper.scala:79)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:495)
[info]   at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1555)
[info]   at org.scalatest.Assertions$class.fail(Assertions.scala:1328)
[info]   at org.scalatest.FunSuite.fail(FunSuite.scala:1555)
[info]   at org.apache.spark.sql.catalyst.expressions.ExpressionEvalHelper$class.generateProject(ExpressionEvalHelper.scala:79)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite.generateProject(CollectionFunctionsSuite.scala:24)
[info]   at org.apache.spark.sql.catalyst.expressions.ExpressionEvalHelper$class.checkEvaluationWithGeneratedMutableProjection(ExpressionEvalHelper.scala:109)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite.checkEvaluationWithGeneratedMutableProjection(CollectionFunctionsSuite.scala:24)
[info]   at org.apache.spark.sql.catalyst.expressions.ExpressionEvalHelper$class.checkEvaluation(ExpressionEvalHelper.scala:44)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite.checkEvaluation(CollectionFunctionsSuite.scala:24)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite$$anonfun$1.apply$mcV$sp(CollectionFunctionsSuite.scala:31)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite$$anonfun$1.apply(CollectionFunctionsSuite.scala:26)
[info]   at org.apache.spark.sql.catalyst.expressions.CollectionFunctionsSuite$$anonfun$1.apply(CollectionFunctionsSuite.scala:26)
[info]   at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:166)
[info]   at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:42)
[info]   at org.scalatest.FunSuiteLike$class.invokeWithFixture$1(FunSuiteLike.scala:163)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.FunSuiteLike$class.runTest(FunSuiteLike.scala:175)
[info]   at org.scalatest.FunSuite.runTest(FunSuite.scala:1555)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
[info]   at scala.collection.immutable.List.foreach(List.scala:318)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:396)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
[info]   at org.scalatest.FunSuiteLike$class.runTests(FunSuiteLike.scala:208)
[info]   at org.scalatest.FunSuite.runTests(FunSuite.scala:1555)
[info]   at org.scalatest.Suite$class.run(Suite.scala:1424)
[info]   at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1555)
[info]   at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
[info]   at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:545)
[info]   at org.scalatest.FunSuiteLike$class.run(FunSuiteLike.scala:212)
[info]   at org.scalatest.FunSuite.run(FunSuite.scala:1555)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:462)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:671)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:294)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:284)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:262)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
```

Not sure why this is, the only difference is removing the pattern matching, which doesn't seem like it should matter since the only accepted input types define `size()`.

tarekauel replied:
```Scala
 override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
    nullSafeCodeGen(ctx, ev, c => s"${ev.primitive} = ($c).size();")
  }
```

`nullSafeCodeGen` allows you to add multiple lines.  `defineCodeGen` expects only the right part of the assignment.

EntilZha replied:
So in that case, what variable does it expect me to store it in (tried returning it, which doesn't work)? What would be the side effect of using `defineCodeGen` and setting nullable to false? It should still take null input and handle it appropriately correct?

EntilZha replied:
Answered my prior question, would be `${ev.primitive} = ...`. So I have this passing tests:

```scala
override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
    nullSafeCodeGen(ctx, ev, c => s"${ev.primitive} = ($c).size();")
  }
```

Look good to push?

EntilZha commented:
On a side note, why is Jenkins saying that there are extra classes being exposed which are not part of the PR diff?

tarekauel commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionFunctionsSuite.scala:
```diff
@@ -0,0 +1,43 @@
+
+    checkEvaluation(Size(m0), 2)
+    checkEvaluation(Size(m1), 0)
+    checkEvaluation(Size(m2), 1)
```
Can you add something like
```Scala
checkEvaluation(Literal.create(null, MapType(StringType, StringType)), null)
checkEvaluation(Literal.create(null, ArrayType(StringType)), null)
```

EntilZha replied:
Done.

Commit:
Added null test case and changed to nullSafeCodeGen

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37869 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37869/consoleFull) for   PR 7462 at commit [`15d4bf1`](https://github.com/apache/spark/commit/15d4bf1281f238f46fdc22c186b423b7ecc56057).

tarekauel commented on python/pyspark/sql/functions.py:
```diff
@@ -795,6 +796,22 @@ def weekofyear(col):
+    >>> from pyspark.sql import Row
+    >>> from pyspark.sql.functions import size
+    >>> df = sqlContext.createDataFrame([Row(data=[1, 2, 3]), Row(data=[1]), Row(data=[])])
+    >>> df.select(size(df.data)).collect()
```
You don't have to import `size` and `Row`.
simple use
```Python
>>> df = sqlContext.createDataFrame([([1, 2, 3],),([1],),([],)], ['data'])
>>> df.select(size(df.data)).collect()
```

EntilZha replied:
Looks like not having these is making the tests fail, so going to re-add it

EntilZha replied:
Nvm, saw your updated code

Commit:
removed imports from python docs

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37870 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37870/consoleFull) for   PR 7462 at commit [`9aea3bb`](https://github.com/apache/spark/commit/9aea3bba05f25e18eaba8168f08d688cd2627490).

SparkQA commented:
[Test build #37869 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37869/console) for   PR 7462 at commit [`15d4bf1`](https://github.com/apache/spark/commit/15d4bf1281f238f46fdc22c186b423b7ecc56057).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Size(child: Expression) extends UnaryExpression with ExpectsInputTypes `

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #37870 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37870/console) for   PR 7462 at commit [`9aea3bb`](https://github.com/apache/spark/commit/9aea3bba05f25e18eaba8168f08d688cd2627490).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Size(child: Expression) extends UnaryExpression with ExpectsInputTypes `

AmplabJenkins commented:
Merged build finished. Test FAILed.

chenghao-intel commented on python/pyspark/sql/functions.py:
```diff
@@ -51,7 +51,8 @@
     'udf',
-    'when']
+    'when',
+    'size']
```
It will be great if we can sort the function in alphabet order.

chenghao-intel commented:
Failed in python test, You can run the python locally with following commands
```
build/sbt -Phive -Phive-thriftserver assembly
python/run-tests --modules=pyspark-sql
```

EntilZha committed:
fixed functions and sorted __all__

EntilZha commented:
@chenghao-intel, last commit should fix it.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37915 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37915/consoleFull) for   PR 7462 at commit [`9a442ae`](https://github.com/apache/spark/commit/9a442aea0dca550ed03a8d87298bb936ae3b215e).

SparkQA commented:
[Test build #37915 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37915/console) for   PR 7462 at commit [`9a442ae`](https://github.com/apache/spark/commit/9a442aea0dca550ed03a8d87298bb936ae3b215e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Size(child: Expression) extends UnaryExpression with ExpectsInputTypes `

AmplabJenkins commented:
Merged build finished. Test PASSed.

chenghao-intel commented on sql/core/src/main/scala/org/apache/spark/sql/functions.scala:
```diff
@@ -2400,6 +2401,25 @@ object functions {
+   * @group collection_funcs
+   * @since 1.5.0
+   */
+  def size(columnName: String): Column = size(Column(columnName))
```
Nit: We just removed the functions with the column name, probably we can leave it for the further updating.

chenghao-intel commented:
LTGM, except a minor issue. @rxin can we merge this?

rxin commented:
Thanks - I've merged this.
